contracts/modules/Liquidation.sol

```diff
@@ -99,61 +119,142 @@ contract Liquidation is BaseLogic {
-            uint borrowAdj = borrowFactor != 0 ? TARGET_HEALTH * CONFIG_FACTOR_SCALE / borrowFactor : MAX_SANE_DEBT_AMOUNT;
-            uint collateralAdj = 1e18 * uint(collateralFactor) / CONFIG_FACTOR_SCALE * 1e18 / (1e18 - liqOpp.discount);
+    function calculateRepayCommon(LiquidationOpportunity memory liqOpp, LiquidationLocals memory liqLocs, uint collateralFactor, uint borrowFactor) private pure {
+        // These factors are first converted into standard 1e18-scale fractions, then adjusted according to TARGET_HEALTH and the discount:
+        uint borrowAdj = borrowFactor != 0 ? TARGET_HEALTH * CONFIG_FACTOR_SCALE / borrowFactor : MAX_SANE_DEBT_AMOUNT;
+        uint collateralAdj = 1e18 * collateralFactor / CONFIG_FACTOR_SCALE * 1e18 / (1e18 - liqOpp.discount);
+        uint liabilityValue = liqLocs.liabilityValue * TARGET_HEALTH / 1e18;
 
+        if (liabilityValue > liqLocs.collateralValue) {
             if (borrowAdj <= collateralAdj) {
                 liqOpp.repay = type(uint).max;
             } else {
-                // liabilityValueTarget >= liabilityValue > collateralValue
-                uint maxRepayInReference = (liabilityValueTarget - collateralValue) * 1e18 / (borrowAdj - collateralAdj);
+                uint maxRepayInReference = (liabilityValue - liqLocs.collateralValue) * 1e18 / (borrowAdj - collateralAdj);
                 liqOpp.repay = maxRepayInReference * 1e18 / liqLocs.underlyingPrice;
             }
         }
+    }
 
+    function boundRepayAndYield(LiquidationOpportunity memory liqOpp, LiquidationLocals memory liqLocs) private pure {
         // Limit repay to current owed
         // This can happen when there are multiple borrows and liquidating this one won't bring the violator back to solvency
 
-        {
-            uint currentOwed = getCurrentOwed(underlyingAssetStorage, underlyingAssetCache, liqLocs.violator);
-            if (liqOpp.repay > currentOwed) liqOpp.repay = currentOwed;
+        if (liqOpp.repay > liqLocs.currentOwed) {
+            liqOpp.repay = liqLocs.currentOwed;
         }
 
         // Limit yield to borrower's available collateral, and reduce repay if necessary
         // This can happen when borrower has multiple collaterals and seizing all of this one won't bring the violator back to solvency
 
         liqOpp.yield = liqOpp.repay * liqOpp.conversionRate / 1e18;
 
-        {
-            uint collateralBalance = balanceToUnderlyingAmount(collateralAssetCache, collateralAssetStorage.users[liqLocs.violator].balance);
-
-            if (collateralBalance < liqOpp.yield) {
-                liqOpp.repay = collateralBalance * 1e18 / liqOpp.conversionRate;
-                liqOpp.yield = collateralBalance;
-            }
+        if (liqLocs.collateralBalance < liqOpp.yield) {
+            liqOpp.repay = liqLocs.collateralBalance * 1e18 / liqOpp.conversionRate;
```

Comment:
maybe I don't seem something, but does `liqOpp.repay` have to be limited here? although not profitable, if the liquidator wants to repay more than the value of the collateral seized, why do we stop them?

Reply:
I think liquidator can repay more by setting `desiredRepay` in `liquidation`, which is then `desiredRepay` in `executeLiquidation`

Reply:
Possibly we'd want to allow more repay, but I think the place to handle it would be in `executeLiquidation`. We can discuss it, but I don't think it's in scope